crafting.py

```diff
@@ -0,0 +1,316 @@
+            self.items = array.array('I', items)
+        else:
+            self.items = array.array('I', [0 for item in items_by_index])
+
+    def __add__(self, other:'State') -> 'State':
+        s = State(self.items)
+        
+        for i in range(len (other.items)) :
+            s.items[i] = s.items[i] + other.items[i]
+        # A. How do we add together the contents of two states?
+        return s
+
+    def __sub__(self, other:'State') -> 'State':
+        s = State(self.items)
+        for i in range(len (other.items)) :
+            s.items[i] = max(s.items[i] - other.items[i],0)
+        # B. How do we subtract one state from another?
+        return s
+
+    def __ge__(self, other: 'State') -> bool:
+        for i in range(len (other.items)) :
+            if self.items[i] < other.items[i] :
+                return False
+        # C. How do we know whether one state (self) contains everything that's inside of another (other)? 
+        return True
+
+    def __lt__(self, other: 'State') -> bool:
+        return not (self >= other)
+
+    def __eq__(self, other) -> bool:
+        return self.items == other.items
+
+    def __hash__(self) -> int:
+        hsh = 5381
+        for s in self.items:
+            hsh = ((hsh << 5) + hsh) + s
+        return hsh
+
+    def __str__(self) -> str:
+        return self.to_dict().__str__()
+
+    def to_dict(self) -> Dict[str, int]:
+        return {items_by_index[idx]: self.items[idx]
+                for idx in range(len(self.items))}
+
+    @classmethod
+    def from_dict(cls, item_dict: Dict[str, int]) -> 'State':
+        return cls([
+            item_dict.get(item,0) for item in items_by_index
+        ])
+
+class Recipe(NamedTuple):    
+    produces: State
+    consumes: State
+    requires: State
+    cost: int
+
+recipes: Dict[str, Recipe] = {}
+for name, rule in Crafting['Recipes'].items():
+    recipes[name] = Recipe(
+        State.from_dict(rule.get('Produces', {})),
+        State.from_dict(rule.get('Consumes', {})),
+        State.from_dict({item: 1 if req else 0
+                         for item, req in rule.get('Requires', {}).items()}),
+        rule['Time']
+    )   
+
+class Proposition(NamedTuple):
+    item: int
+    at_least: int
+
+def state_propositions(state: State) -> Set[Proposition]:
+    propositions: Set[Proposition] = set()
+    
+    for i in range(len (state.items)) :
+        for k in range(1,state.items[i]+1):
+            propositions.add(Proposition(i,k))
+     
+    return propositions
+
+
+def recipe_to_propositions(recipe: Recipe) -> Set[Proposition]:
+    propositions: Set[Proposition] = set()
+    # G. Do something with recipe.consumes, recipe.produces, and recipe.requires.
+    # Emit, for this recipe, all the propositions entailed by the preconditions and the _minimal_ set of propositions embodied in the preconditions (i.e., don't need to output wood >= 2, wood >= 1, wood >= 0 if the recipe creates 2 wood.)
```

Comment:
This comment had a typo; the "produces" part should only be a minimal set, i.e. just the propositions for the exact amount the recipe produces, to minimize the number of proposition combinations you need to consider.
So the minimal set of propositions embodied in the *postconditions*.